File: app/src/main/java/com/hotelnow/activity/SettingActivity.java
package com.hotelnow.activity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import com.hotelnow.R;
import com.hotelnow.dialog.DialogAgreeUser;
import com.hotelnow.dialog.DialogAlert;
import com.hotelnow.utils.Api;
import com.hotelnow.utils.CONFIG;
import com.hotelnow.utils.HotelnowApplication;
import com.hotelnow.utils.LogUtil;
import com.hotelnow.utils.OnSingleClickListener;
import com.hotelnow.utils.Util;
import com.koushikdutta.ion.Ion;
import com.squareup.okhttp.Response;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import java.util.Map;

public class SettingActivity extends Activity {

    private SharedPreferences _preferences;
    private String cookie = "";
    private DialogAgreeUser dialogAgreeUser;
    private boolean Sel_check = false;
    private ImageView my_banner;
    private DialogAlert dialogAlert;
    private String url_link = "";
    private TextView tv_alarm_setting;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_setting);

        _preferences = PreferenceManager.getDefaultSharedPreferences(this);
        cookie = _preferences.getString("userid", null);
        my_banner = (ImageView) findViewById(R.id.my_banner);
        tv_alarm_setting = (TextView) findViewById(R.id.tv_alarm_setting);
        if(cookie == null){
            tv_alarm_setting.setText("할인 혜택 알림 받기(Push)");
            findViewById(R.id.btn_retire).setVisibility(View.GONE);
            findViewById(R.id.retire_line).setVisibility(View.GONE);
        }
        else{
            tv_alarm_setting.setText("할인 혜택 알림 받기(SMS,E-mail,Push)");
            findViewById(R.id.btn_retire).setVisibility(View.VISIBLE);
            findViewById(R.id.retire_line).setVisibility(View.VISIBLE);
        }
        // 알림 받기
        findViewById(R.id.btn_push).setOnClickListener(new OnSingleClickListener() {
            @Override
            public void onSingleClick(View v) {
                Intent intent = new Intent(SettingActivity.this, SettingAlarmActivity.class);
                startActivity(intent);
            }
        });

        // 개인정보 수집
        findViewById(R.id.btn_agree1).setOnClickListener(new OnSingleClickListener() {
            @Override
            public void onSingleClick(View v) {
                setUserBenefit(true);
            }
        });

        // 위치서비스
        findViewById(R.id.btn_agree2).setOnClickListener(new OnSingleClickListener() {
            @Override
            public void onSingleClick(View v) {
                Intent intent = new Intent(SettingActivity.this, LocationAgreeActivity.class);
                startActivity(intent);
            }
        });

        // 탈퇴하기
        findViewById(R.id.btn_retire).setOnClickListener(new OnSingleClickListener() {
            @Override
            public void onSingleClick(View v) {
                Intent intent = new Intent(SettingActivity.this, PwCheckActivity.class);
                startActivityForResult(intent, 999);
            }
        });

        findViewById(R.id.title_back).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setResult(91);
                finish();
            }
        });

        setUserBenefit(false);
    }

    private void setUserBenefit(final boolean flag){
        findViewById(R.id.wrapper).setVisibility(View.VISIBLE);
        String url = CONFIG.maketing_agree;
        String uuid = Util.getAndroidId(this);

        if(uuid != null && !TextUtils.isEmpty(uuid)){
            url += "?uuid="+uuid;
        }
        url +="&marketing_use";

        Api.get(url, new Api.HttpCallback() {
            @Override
            public void onFailure(Response response, Exception throwable) {
                findViewById(R.id.wrapper).setVisibility(View.GONE);
                Toast.makeText(SettingActivity.this, getString(R.string.error_connect_problem), Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onSuccess(Map<String, String> headers, String body) {
                try {
                    JSONObject obj = new JSONObject(body);

                    if (!obj.getString("result").equals("success")) {
                        Toast.makeText(SettingActivity.this, obj.getString("msg"), Toast.LENGTH_SHORT).show();
                        findViewById(R.id.wrapper).setVisibility(View.GONE);
                        return;
                    }
                    String agreed_yn = "N";
                    if(obj.has("marketing_use")){
                        agreed_yn = obj.getJSONObject("marketing_use").getString("agreed_yn");
                    }

                    if(flag) {
                        setAgreedPopup(agreed_yn);
                    }else{
                        if(obj.has("event_banners")){
                            findViewById(R.id.layout_banner).setVisibility(View.VISIBLE);
                            Ion.with(my_banner).load(obj.getJSONArray("event_banners").getJSONObject(0).getString("image"));
                            final String id = obj.getJSONArray("event_banners").getJSONObject(0).getString("event_id");
                            final String evt_type = obj.getJSONArray("event_banners").getJSONObject(0).getString("evt_type");
                            final String link = obj.getJSONArray("event_banners").getJSONObject(0).getString("link");
                            final String title = obj.getJSONArray("event_banners").getJSONObject(0).getString("title");
                            findViewById(R.id.layout_banner).setOnClickListener(new OnSingleClickListener() {
                                @Override
                                public void onSingleClick(View v) {
                                    setEventCheck(id, evt_type, link, title);
                                }
                            });
                        }
                        else{
                            findViewById(R.id.layout_banner).setVisibility(View.GONE);
                        }
                    }
                    findViewById(R.id.wrapper).setVisibility(View.GONE);

                } catch (Exception e) {
                    findViewById(R.id.wrapper).setVisibility(View.GONE);
                }
            }
        });
    }

    private void setMaketing(boolean ischeck) {
        // 푸시 수신 상태값 저장
        String regId = _preferences.getString("gcm_registration_id", null);

        LogUtil.e("xxxxx", regId);
        if (regId != null) {
            setMaketingSend(this, regId, ischeck);
        }
    }

    // GCM TOKEN
    public void setMaketingSend(final Context context, String regId, final Boolean flag) {
        String androidId = Util.getAndroidId(context);

        JSONObject paramObj = new JSONObject();

        try {
            paramObj.put("os", "a");
            paramObj.put("uuid", androidId);
            paramObj.put("push_token", regId);
            paramObj.put("ver", Util.getAppVersionName(context));
            paramObj.put("marketing_use", (flag == true) ? "Y" : "N");

        } catch (JSONException e) {; }

        Api.post(CONFIG.maketing_agree_change, paramObj.toString(), new Api.HttpCallback() {
            @Override
            public void onFailure(Response response, Exception e) {

            }

            @Override
            public void onSuccess(Map<String, String> headers, String body) {
                try {
                    JSONObject obj = new JSONObject(body);

                    if (!obj.getString("result").equals("success")) {
                        Toast.makeText(HotelnowApplication.getAppContext(), obj.getString("msg"), Toast.LENGTH_SHORT).show();
                        return;
                    }

                } catch (Exception e) {
                }
            }
        });
    }

    private void setAgreedPopup(String value){
        dialogAgreeUser = new DialogAgreeUser(SettingActivity.this,
                new OnSingleClickListener() {
                    @Override
                    public void onSingleClick(View v) {
                        // api 호출
                        setMaketing(Sel_check);
                        dialogAgreeUser.dismiss();
                    }
                },
                new OnSingleClickListener() {
                    @Override
                    public void onSingleClick(View v) {
                        dialogAgreeUser.dismiss();
                    }
                }, value, new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                Sel_check = isChecked;
            }
        }, cookie == null ? false : true);
        dialogAgreeUser.setCancelable(false);
        dialogAgreeUser.show();
    }

    private void setEventCheck(String id, String type, String link, String title){

        String[] arr = link.split("hotelnowevent://");
        String frontMethod = "";
        String frontTitle = "";
        String frontEvtId = "";
        String method = "";

        if (arr.length > 1) {
            frontMethod = arr[1];
            frontMethod = Util.stringToHTMLString(frontMethod);
            frontTitle = title != "" ? title : "무료 숙박 이벤트";
        }
        if (!type.equals("a")) {
            frontEvtId = id;
        } else {
            frontEvtId = Util.getFrontThemeId(link);
        }

        if (type.equals("a") && !type.equals("")) {
            try {
                JSONObject obj = new JSONObject(frontMethod);
                method = obj.getString("method");
                if(obj.has("param")) {
                    url_link = obj.getString("param");
                }

                if (method.equals("move_near")) {
                    int fDayLimit = _preferences.getInt("future_day_limit", 180);
                    String checkurl = CONFIG.checkinDateUrl + "/" + url_link + "/" + fDayLimit;

                    Api.get(checkurl, new Api.HttpCallback() {
                        @Override
                        public void onFailure(Response response, Exception e) {
                            Toast.makeText(SettingActivity.this, getString(R.string.error_connect_problem), Toast.LENGTH_SHORT).show();
                            return;
                        }

                        @Override
                        public void onSuccess(Map<String, String> headers, String body) {
                            try {
                                JSONObject obj = new JSONObject(body);
                                JSONArray aobj = obj.getJSONArray("data");

                                if (aobj.length() == 0) {
                                    dialogAlert = new DialogAlert(
                                            getString(R.string.alert_notice),
                                            "해당 숙소는 현재 예약 가능한 객실이 없습니다.",
                                            SettingActivity.this,
                                            new View.OnClickListener() {
                                                @Override
                                                public void onClick(View v) {
                                                    dialogAlert.dismiss();
                                                }
                                            });
                                    dialogAlert.setCancelable(false);
                                    dialogAlert.show();
                                    return;
                                }

                                String checkin = aobj.getString(0);
                                String checkout = Util.getNextDateStr(checkin);

                                Intent intent = new Intent(SettingActivity.this, DetailHotelActivity.class);
                                intent.putExtra("hid", url_link);
                                intent.putExtra("evt", "N");
                                intent.putExtra("sdate", checkin);
                                intent.putExtra("edate", checkout);

                                startActivity(intent);

                            } catch (Exception e) {
                                // Log.e(CONFIG.TAG, e.toString());
                                Toast.makeText(SettingActivity.this, getString(R.string.error_connect_problem), Toast.LENGTH_SHORT).show();
                                return;
                            }

                        }
                    });
                } else if (method.equals("move_theme")) {
                    Intent intent = new Intent(SettingActivity.this, ThemeSpecialHotelActivity.class);
                    intent.putExtra("tid", url_link);

                    startActivity(intent);

                } else if (method.equals("move_theme_ticket")) {
                    Intent intent = new Intent(SettingActivity.this, ThemeSpecialActivityActivity.class);
                    intent.putExtra("tid", url_link);

                    startActivity(intent);

                } else if (method.equals("move_ticket_detail")) {
                    Intent intent = new Intent(SettingActivity.this, DetailActivityActivity.class);
                    intent.putExtra("tid", url_link);

                    startActivity(intent);

                } else if (method.equals("outer_link")) {
                    if (url_link.contains("hotelnow")) {
                        frontTitle = title != "" ? title : "무료 숙박 이벤트";
                        Intent intent = new Intent(SettingActivity.this, WebviewActivity.class);
                        intent.putExtra("url", url_link);
                        intent.putExtra("title", frontTitle);
                        startActivity(intent);

                    } else {
                        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url_link));
                        startActivity(intent);
                    }
                } else if (method.equals("move_privatedeal_all")){
                    Intent intent = new Intent(SettingActivity.this, PrivateDaelAllActivity.class);
                    startActivity(intent);
                }
            } catch (Exception e) {
            }
        } else {
            frontTitle = title != "" ? title : "무료 숙박 이벤트";
            Intent intentEvt = new Intent(SettingActivity.this, EventActivity.class);
            intentEvt.putExtra("idx", Integer.valueOf(frontEvtId));
            intentEvt.putExtra("title", frontTitle);
            startActivity(intentEvt);
        }
    }

    @Override
    public void onBackPressed() {
        setResult(91);
        finish();
        super.onBackPressed();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(requestCode == 999)
        {
            if(resultCode == 888){
                finish();
            }
            else if(resultCode == 999){
                setResult(999);
                finish();
            }
        }
    }
}

File: app/src/main/java/com/hotelnow/activity/SettingActivity.kt
package com.hotelnow.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.preference.PreferenceManager
import android.text.TextUtils
import android.view.View
import android.widget.CompoundButton
import android.widget.Toast
import com.hotelnow.R
import com.hotelnow.dialog.DialogAgreeUser
import com.hotelnow.dialog.DialogAlert
import com.hotelnow.utils.*
import com.koushikdutta.ion.Ion
import com.squareup.okhttp.Response
import kotlinx.android.synthetic.main.activity_setting.*
import org.json.JSONException
import org.json.JSONObject

class SettingActivity : Activity() {

    private var _preferences: SharedPreferences? = null
    private var cookie: String? = ""
    private var dialogAgreeUser: DialogAgreeUser? = null
    private var Sel_check: Boolean = false
    private var dialogAlert: DialogAlert? = null
    private var url_link: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_setting)

        _preferences = PreferenceManager.getDefaultSharedPreferences(this@SettingActivity)
        _preferences?.let {
                    cookie = it.getString("userid", null)
        }
        cookie?.let {
            tv_alarm_setting.text = "할인 혜택 알림 받기(Push)"
            btn_retire.visibility = View.GONE
            retire_line.visibility = View.GONE
        }.let {
            tv_alarm_setting.text = "할인 혜택 알림 받기(SMS,E-mail,Push)"
            btn_retire.visibility = View.VISIBLE
            retire_line.visibility = View.VISIBLE
        }

        btn_push.setOnClickListener(object : OnSingleClickListener() {
            override fun onSingleClick(v: View) {
                val intent = Intent(this@SettingActivity, SettingAlarmActivity::class.java)
                startActivity(intent)
            }
        })

        // 개인정보 수집
        btn_agree1.setOnClickListener(object : OnSingleClickListener() {
            override fun onSingleClick(v: View) {
                setUserBenefit(true)
            }
        })

        // 위치서비스
        btn_agree2.setOnClickListener(object : OnSingleClickListener() {
            override fun onSingleClick(v: View) {
                val intent = Intent(this@SettingActivity, LocationAgreeActivity::class.java)
                startActivity(intent)
            }
        })

        // 탈퇴하기
        btn_retire.setOnClickListener(object : OnSingleClickListener() {
            override fun onSingleClick(v: View) {
                val intent = Intent(this@SettingActivity, PwCheckActivity::class.java)
                startActivityForResult(intent, 999)
            }
        })

        title_back.setOnClickListener() {
            setResult(91)
            finish()
        }

        setUserBenefit(false)
    }

    private fun setUserBenefit(flag: Boolean){
        wrapper.visibility = View.VISIBLE
        var url = CONFIG.maketing_agree
        val uuid = Util.getAndroidId(this)

        if (uuid != null && !TextUtils.isEmpty(uuid)) {
            url += "?uuid=$uuid"
        }
        url += "&marketing_use"

        Api.get(url, object : Api.HttpCallback {
            override fun onFailure(response: Response, throwable: Exception) {
                wrapper.visibility = View.GONE
                Toast.makeText(this@SettingActivity, getString(R.string.error_connect_problem), Toast.LENGTH_SHORT).show()
            }

            override fun onSuccess(headers: Map<String, String>, body: String) {
                try {
                    val obj = JSONObject(body)

                    if (obj.getString("result") != "success") {
                        Toast.makeText(this@SettingActivity, obj.getString("msg"), Toast.LENGTH_SHORT).show()
                        wrapper.visibility = View.GONE
                        return
                    }
                    var agreed_yn = "N"
                    if (obj.has("marketing_use")) {
                        agreed_yn = obj.getJSONObject("marketing_use").getString("agreed_yn")
                        Sel_check = if (obj.getJSONObject("marketing_use").getString("agreed_yn") == "Y") true else false
                    }

                    if (flag) {
                        setAgreedPopup(agreed_yn)
                    } else {
                        if (obj.has("event_banners") && obj.getJSONArray("event_banners").length() > 0) {
                            findViewById<View>(R.id.layout_banner).visibility = View.VISIBLE
                            Ion.with(my_banner).load(obj.getJSONArray("event_banners").getJSONObject(0).getString("image"))
                            val id = obj.getJSONArray("event_banners").getJSONObject(0).getString("event_id")
                            val evt_type = obj.getJSONArray("event_banners").getJSONObject(0).getString("evt_type")
                            val link = obj.getJSONArray("event_banners").getJSONObject(0).getString("link")
                            val title = obj.getJSONArray("event_banners").getJSONObject(0).getString("title")
                            layout_banner.setOnClickListener(object : OnSingleClickListener() {
                                override fun onSingleClick(v: View) {
                                    setEventCheck(id, evt_type, link, title)
                                }
                            })
                        } else {
                            layout_banner.visibility = View.GONE
                        }
                    }
                    wrapper.visibility = View.GONE

                } catch (e: Exception) {
                    wrapper.visibility = View.GONE
                }

            }
        })
    }

    private fun setAgreedPopup(value: String){
        dialogAgreeUser = DialogAgreeUser(this@SettingActivity,
                object : OnSingleClickListener() {
                    override fun onSingleClick(v: View) {
                        // api 호출
                        setMaketing(Sel_check)
                        dialogAgreeUser?.dismiss()
                    }
                },
                object : OnSingleClickListener() {
                    override fun onSingleClick(v: View) {
                        dialogAgreeUser?.dismiss()
                    }
                }, value, CompoundButton.OnCheckedChangeListener { _: CompoundButton, isChecked -> Sel_check = isChecked }, if (cookie == null) false else true)

        dialogAgreeUser?.let {
            it.setCancelable(false)
            it.show()
        }
    }

    private fun setMaketing(ischeck: Boolean) {
        // 푸시 수신 상태값 저장
        var regId: String? = null
        _preferences?.let{
            regId = it.getString("gcm_registration_id", null)
        }

        regId?.let {
            LogUtil.e("xxxxx", it)
            setMaketingSend(this, it, ischeck)
        }.let {
            setMaketingSend(this, "", ischeck)
        }
    }

    // GCM TOKEN
    fun setMaketingSend(context: Context, regId: String, flag: Boolean?) {
        val androidId = Util.getAndroidId(context)

        val paramObj = JSONObject()

        try {
            paramObj.apply {
                put("os", "a")
                put("uuid", androidId)
                put("push_token", regId)
                put("ver", Util.getAppVersionName(context))
                put("marketing_use", if (flag == true) "Y" else "N")
            }

        } catch (e: JSONException) {
        }

        Api.post(CONFIG.maketing_agree_change, paramObj.toString(), object : Api.HttpCallback {
            override fun onFailure(response: Response, e: Exception) {

            }

            override fun onSuccess(headers: Map<String, String>, body: String) {
                try {
                    val obj = JSONObject(body)

                    if (obj.getString("result") != "success") {
                        Toast.makeText(HotelnowApplication.getAppContext(), obj.getString("msg"), Toast.LENGTH_SHORT).show()
                        return
                    }

                } catch (e: Exception) {
                }

            }
        })
    }

    private fun setEventCheck(id: String, type: String, link: String, title: String) {

        val arr = link.split("hotelnowevent://".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var frontMethod: String? = ""
        var frontTitle: String?
        var frontEvtId: String?
        var method: String?

        if (arr.size > 1) {
            frontMethod = arr[1]
            frontMethod = Util.stringToHTMLString(frontMethod)
            frontTitle = if (title !== "") title else "무료 숙박 이벤트"
        }
        if (type != "a") {
            frontEvtId = id
        } else {
            frontEvtId = Util.getFrontThemeId(link)
        }

        if (type == "a" && type != "") {
            try {
                val obj = JSONObject(frontMethod)
                method = obj.getString("method")
                if (obj.has("param")) {
                    url_link = obj.getString("param")
                }

                if (method == "move_near") {
                    var fDayLimit = 180
                    _preferences?.let { fDayLimit = it.getInt("future_day_limit", 180) }
                    val checkurl = CONFIG.checkinDateUrl + "/" + url_link + "/" + fDayLimit

                    Api.get(checkurl, object : Api.HttpCallback {
                        override fun onFailure(response: Response, e: Exception) {
                            Toast.makeText(this@SettingActivity, getString(R.string.error_connect_problem), Toast.LENGTH_SHORT).show()
                            return
                        }

                        override fun onSuccess(headers: Map<String, String>, body: String) {
                            try {
                                val _obj = JSONObject(body)
                                val aobj = _obj.getJSONArray("data")

                                if (aobj.length() == 0) {
                                    dialogAlert = DialogAlert(
                                            getString(R.string.alert_notice),
                                            "해당 숙소는 현재 예약 가능한 객실이 없습니다.",
                                            this@SettingActivity,
                                            View.OnClickListener { dialogAlert?.dismiss() })
                                    dialogAlert?.let {
                                        it.setCancelable(false)
                                        it.show()
                                    }
                                    return
                                }

                                val checkin = aobj.getString(0)
                                val checkout = Util.getNextDateStr(checkin)

                                val intent = Intent(this@SettingActivity, DetailHotelActivity::class.java)
                                intent.apply {
                                    putExtra("hid", url_link)
                                    putExtra("evt", "N")
                                    putExtra("sdate", checkin)
                                    putExtra("edate", checkout)
                                }
                                startActivity(intent)

                            } catch (e: Exception) {
                                // Log.e(CONFIG.TAG, e.toString());
                                Toast.makeText(this@SettingActivity, getString(R.string.error_connect_problem), Toast.LENGTH_SHORT).show()
                                return
                            }

                        }
                    })
                } else if (method == "move_theme") {
                    val intent = Intent(this@SettingActivity, ThemeSpecialHotelActivity::class.java)
                    intent.apply {
                        putExtra("tid", url_link)
                    }
                    startActivity(intent)

                } else if (method == "move_theme_ticket") {
                    val intent = Intent(this@SettingActivity, ThemeSpecialActivityActivity::class.java)
                    intent.apply {
                        putExtra("tid", url_link)
                    }
                    startActivity(intent)
                } else if (method == "move_ticket_detail") {
                    val intent = Intent(this@SettingActivity, DetailActivityActivity::class.java)
                    intent.apply {
                        putExtra("tid", url_link)
                    }
                    startActivity(intent)
                } else if (method == "outer_link") {
                    if (url_link.contains("hotelnow")) {
                        frontTitle = if (title !== "") title else "무료 숙박 이벤트"
                        val intent = Intent(this@SettingActivity, WebviewActivity::class.java)
                        intent.apply {
                            putExtra("url", url_link)
                            putExtra("title", frontTitle)
                        }
                        startActivity(intent)
                    } else {
                        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url_link))
                        startActivity(intent)
                    }
                } else if (method == "move_privatedeal_all") {
                    val intent = Intent(this@SettingActivity, PrivateDaelAllActivity::class.java)
                    startActivity(intent)
                }
            } catch (e: Exception) {
            }

        } else {
            frontTitle = if (title !== "") title else "무료 숙박 이벤트"
            val intentEvt = Intent(this@SettingActivity, EventActivity::class.java)
            intentEvt.apply {
                frontEvtId?.let { intentEvt.putExtra("idx", Integer.valueOf(frontEvtId)) }
                putExtra("title", frontTitle)
            }
            startActivity(intentEvt)
        }
    }

    override fun onBackPressed() {
        setResult(91)
        finish()
        super.onBackPressed()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == 999) {
            if (resultCode == 888) {
                finish()
            } else if (resultCode == 999) {
                setResult(999)
                finish()
            }
        }
    }
}